File: battleship/static/javascript/ships.js
'use strict';
var fleet=require('./fleet.js');

// Config settings 
let ship_config = {
    aircraftCarrier : {
        size : 5,
        id : 'aircraftCarrier',
        color : 'Crimson',
        clickClass : 'acclicked',
        label : 'Aircraft Carrier',
	mask : 31,
    },
    battleship : {
        size : 4,
        id : 'battleship',
        color:'DarkGreen',
        clickClass : 'bsclicked',
        label : 'Battleship',
	mask: 15,
    },
    destroyer : {
        size : 3,
        id : 'destroyer',
        color:'CadetBlue',
        clickClass : 'declicked',
        label : 'Destroyer',
	mask: 7,
    },
    submarine  : {
        size : 3,
        id : 'submarine',
        color:'DarkRed',
        clickClass : 'suclicked',
        label : 'Submarine',
	mask : 7,
    },
    patrolBoat : {
        size : 2,
        id : 'patrolBoat',
        color:'Gold',
        clickClass : 'pbclicked',
        label : 'Patrol Boat',
	mask: 3,
    },
};

let hitCounter = {
    aircraftCarrier : 0,
    battleship : 0,
    destroyer : 0,
    submarine  : 0,
    patrolBoat : 0
};

let sunkCounter; // Tracks which boats have been sunk

// Values for determining bit values when a boat sinks
const airCraftCarrier = 1;
const battleship = 2;
const destroyer = 4;
const submarine = 8;
const patrolBoat = 16;

let setHitCounter = function (type, bit) {
	hitCounter[type] = ship_config[type].mask^(bit*bit);
	if (hitCounter[type] == ship_config[type].mask) { // I don't know if this is correct but the idea is check to see if the ship is sunk and flag it if need be
		setSunkCounter(type);
	}
}

let setSunkCounter = function (type) {
	sunkCounter = sunkCounter^type;
}

let getHitCounter = function (type){
	return hitCounter[type];
}

let getSunkCounter = function(){
	return sunkCounter;
}

// Ship constructor - shipyard???
function _ship(size, id, color, clickClass, label) {
        this.size        = size;
        this.id          = id;
        this.color       = color;
        this.clickClass  = clickClass;
        this.label       = label;

        return (this);
}

let ships={};

/*
 * The ship object holds the current orientation of the ship and the start coordinate (topmost/leftmost). When
 * there is a change to the ship the master matrix needs to be updated. An event will be triggered when there is
 * a coordinate change. This listener will update the master matrix. Calls to check location (move validtion, 
 * check if hit, etc.) will be made against the master matrix.
 */
/*
let shipIint = function(){
    addEventListener('shipMove',()) }

}
*/
// Public function to initially create ships object
let buildShips = function (){
    for (let s in ship_config){
        ships[s] = {size: ship_config[s].size, 
		    type: ship_config[s].id,
	            color: ship_config[s].color,
		    clickClass: ship_config[s].clickClass,
		    label: ship_config[s].label
	           };
    }
return ships;
}

let buildShip = function(type){
        ships[type] = _ship(ship_config[type].size, ship_config[type].id, ship_config[type].color, ship_config[type].clickClass, ship_config[type].label);
	return ships;
}

// Set value in ship object. 
let setShip = function(type, key, value){
        if (type && ships[type] && key) { // only attempt an update if there is a legit ship type and a key
            ships[type].key = value;
   }
}

// Return ship object if no type given otherwise return object containing just requested ship
let getShip = function (type){
    if(type){
        return ships[type];
    } else {
        return ships;
    }
}

// Private function to randomly determine ship's orientation along the X-axis or Y-axis. Only used when plotting ships for the first time.
function _getStartCoordinate(size){
    const start_orientation=Math.floor(Math.random()*10) > 5 ? 'x' : 'y';
    const start_x = start_orientation == 'x' ? _getRandomCoordinate(size) : _getRandomCoordinate(0);
    const start_y = start_orientation == 'y' ? _getRandomCoordinate(size) : _getRandomCoordinate(0);

    return {coordinate: start_x + '_' + start_y, orientation: start_orientation};
}

// Take ship size and orientation into account when determining the start range value. ex. don't
// let an aircraft carrier with an orientation of 'X' start at row 7 because it will max out over the
// grid size.
function _getRandomCoordinate(offset){
    const MAX_COORD = 10;
    return Math.floor(Math.random()*(MAX_COORD - offset));

}

// FIXME Does fleet.ghostShip do this now?
// Build an array of coordinates for a ship based on it's orientation, intended start point and size
let _shipString = function(s) {
	const o = s.orientation;
	const st = s.start_coordinate;
	let r = new Array;
        let t_pieces = st.split('_');
	const i = o == 'x' ? 0 : 1;

	for (let j=0; j < s.size;j++) {
		t_pieces[i] = t_pieces[i]+1;
		r.push (t_pieces[0] + '_' + t_pieces[1]);
	}
	return r;
}


/*
 * placeShips - Initial placement of ships on the board
 */
let placeShips = function placeShips(){
        /* Randomly place ships on the grid. In order do this each ship must:
	 *   * Pick an orientation
	 *   * Pick a starting coordinate
	 *   * Validate that the coordinate is valid (does not run OOB, does not cross any other ship, etc.)
	 *   * If valid:
	 *   	* Save start coord and orientation as part of ship object
	 *   	* Plot ship on master matrix
	 */
	let shipList = getShip();
        for (var ship in shipList) {
            
            let start = _getStartCoordinate(shipList[ship].size); 
	    let ship_string = fleet.ghostShip(shipList[ship].type, start.coordinate, start.orientation);
	    shipList[ship].orientation = start.orientation;

            while (!fleet.validateShip(ship_string)) {
                start = _getStartCoordinate(shipList[ship].size); 
		shipList[ship].orientation = start.orientation;
		ship_string = fleet.ghostShip(shipList[ship].type, start.coordinate, start.orientation);
		}

            fleet.setFleet(start.orientation,
                       shipList[ship].type,
                       shipList[ship].size,
                       start.coordinate);
            }
};


module.exports = {
    buildShips: buildShips,
    buildShip: buildShip,
    getShip: getShip,
    setShip: setShip,
    placeShips: placeShips,
    setHitCounter: setHitCounter
}

File: BotPlayer/battleship/ships.py
from random import *
class Ships:
    def __init__(self):
        #   Config settings 
        self.ship_config = {
            'aircraftCarrier': {'size': 5, 'id': 'aircraftCarrier', 'label' : 'Aircraft Carrier', 'mask' : 31},
            'battleship':      {'size': 4, 'id' : 'battleship', 'label' : 'Battleship', 'mask': 15, },
            'destroyer':       {'size': 3, 'id' : 'destroyer', 'label' : 'Destroyer', 'mask': 7, },
            'submarine':       {'size': 3, 'id' : 'submarine', 'label' : 'Submarine', 'mask' : 7, },
            'patrolBoat':      {'size': 2, 'id' : 'patrolBoat', 'label' : 'Patrol Boat', 'mask': 3, },
        }

        self.hitCounter = {'aircraftCarrier' : 0, 'battleship' : 0, 'destroyer' : 0, 'submarine'  : 0, 'patrolBoat' : 0 }

        self.sunkCounter = {} # Tracks which boats have been sunk

        # Values for determining bit values when a boat sinks
        self.airCraftCarrier = 1,
        self.battleship = 2,
        self.destroyer = 4,
        self.submarine = 8,
        self.patrolBoat = 16,

    def setHitCounter (self, type, bit):
        ships.hitCounter[type] = ships.ship_config[type].mask^(bit*bit)
        if ships.hitCounter[type] == ships.ship_config[type].mask: # I don't know if this is correct but the idea is check to see if the ship is sunk and flag it if need be
            ships.setSunkCounter(type)

    def setSunkCounter(self, type):
        ships.sunkCounter = ships.sunkCounter^type;

    def getHitCounter (self, type):
        return ships.hitCounter[type]

    def getSunkCounter (self):
        return ships.sunkCounter
#
#    # Ship constructor - shipyard??? __init ???
#    ship: function(size, id, color, clickClass, label) {
#        this.size        = size;
#        this.id          = id;
#        this.color       = color;
#        this.clickClass  = clickClass;
#        this.label       = label;
#
#        return (this);
#    },

    # The ship object holds the current orientation of the ship and the start coordinate (topmost/leftmost). When
    # there is a change to the ship the master matrix needs to be updated. An event will be triggered when there is
    # a coordinate change. This listener will update the master matrix. Calls to check location (move validtion, 
    # check if hit, etc.) will be made against the master matrix.
    # Public function to initially create ships object
    def buildShips(self):
        for s in ships.ship_config:
            ships[s] = {'size': ships.ship_config[s].size, 'type': ships.ship_config[s].id, 'label': ships.ship_config[s].label}
        return ships

    def buildShip (self, type):
        ships[type] = ships.ship(ships.ship_config[type].size, ships.ship_config[type].id, ships.ship_config[type].label)
        return ships;

    # Set value in ship object. 
    def setShip (self, type, key, value):
        if type and ships[type] and key: # only attempt an update if there is a legit ship type and a key
            ships[type].key = value;

    # Return ship object if no type given otherwise return object containing just requested ship
    def getShip(self, type):
        if type:
            return ships[type]
        else:
            return ships.ship_config

    # Private function to randomly determine ship's orientation along the X-axis or Y-axis. Only used when plotting ships for the first time.
    def getStartCoordinate(self, size):
	#TODO const = javascript immutable...translate
        start_orientation=('x', 'y')[randint(1,10) > 5]
        start_x = (ships.getRandomCoordinate(size), ships.getRandomCoordinate(0))[start_orientation == 'x'] 
        start_y = (ships.getRandomCoordinate(size), ships.getRandomCoordinate(0))[start_orientation == 'y']

        return {coordinate: start_x + '_' + start_y, orientation: start_orientation}

    # Take ship size and orientation into account when determining the start range value. ex. don't
    # let an aircraft carrier with an orientation of 'X' start at row 7 because it will max out over the
    # grid size.
    def getRandomCoordinate (self, offset):
        MAX_COORD = 10
        return randint(0,(MAX_COORD - offset))

    # 
    # placeShips - Initial placement of ships on the board
    # 
    def placeShips(self):
         # Randomly place ships on the grid. In order do this each ship must:
         #   * Pick an orientation
         #   * Pick a starting coordinate
         #   * Validate that the coordinate is valid (does not run OOB, does not cross any other ship, etc.)
         #   * If valid:
         #       * Save start coord and orientation as part of ship object
         #       * Plot ship on master matrix
         #
        shipList = ships.getShip()
        for ship in shipList:
            start = ships.getStartCoordinate(shipList[ship].size)
            ship_string = fleet.ghostShip(ship, start.coordinate, start.orientation)
            shipList[ship].orientation = start.orientation

            while not fleet.validateShip(ship_string):
                start = ships.getStartCoordinate(shipList[ship].size)
                shipList[ship].orientation = start.orientation
                ship_string = fleet.ghostShip(ship, start.coordinate, start.orientation)

            fleet.setFleet(start.orientation, ship, shipList[ship].size, start.coordinate)


